File: Chessko/Views/StepGameView.swift
import SwiftUI

// MARK: - Step Game View (pokretac `game` koraka Puta)
//
// Partija protiv racunara, sa tezinom i (opciono) startnom pozicijom iz koraka.
// Ceo tok partije — tapovi, legalnost, AI, zvuk, haptika, kraj partije — vec
// zivi u `GameViewModel`-u i deli se sa tabom Igra; ovde se dodaje samo ono sto
// korak ima a slobodna partija nema: zavrsetak KORAKA.
//
// Tri odluke koje ovaj ekran nosi:
//
// 1. SOPSTVENI `GameViewModel`, sa sopstvenim slotom u `UserDefaults`-u
//    (`GameViewModel.stepSaveKey(_:)`). Model tabu Igra cuva celu partiju pod
//    jednim kljucem posle SVAKOG poteza; deljen primerak (ili zajednicki kljuc)
//    bi ulaskom u korak ucitao partiju koju korisnik ima u toku na tabu Igra i
//    prvim potezom je pregazio. Isti razlog zbog kojeg `StepPracticeView` drzi
//    svoj `PuzzleViewModel`.
// 2. Korak se zavrsava kad partija dodje do KRAJA, bez obzira na ishod — spec
//    kaze „partija odigrana do kraja", ne „pobeda". Predaja je takodje kraj.
// 3. Ekran NE nudi „Nova igra" ni izbor boje. To su kontrole slobodne partije;
//    ovde bi bile nacin da se korak zavrsi bez partije.

struct StepGameView: View {

    let stepId: String
    /// Tezina dolazi kao `GameDifficulty`, ne kao string iz kurikuluma: mesto
    /// koje sme da promasi mapiranje je `PathView.route(for:)`, jedno jedino.
    let difficulty: GameDifficulty
    let startFEN: String?

    /// Model se pravi u `onAppear`, a ne kao pocetna vrednost `@State`-a:
    /// svako kreiranje `GameViewModel`-a cita `UserDefaults`, budi Stockfish i
    /// pokrece racunanje evaluacije, a pocetna vrednost `@State`-a se izracuna
    /// pri SVAKOM kreiranju ove strukture — a ekran Puta se prekrsti na svaku
    /// promenu napretka. Ovako postoji tacno jedan primerak.
    @State private var viewModel: GameViewModel?

    /// Korak koji jos ceka upis. `nil` znaci „vec upisan": zavrsetak se
    /// okida TACNO jednom po ulasku, a ne na svako ponovno izracunavanje
    /// `isGameOver`-a.
    @State private var pendingStepId: String?

    @State private var showResignConfirm = false

    @Environment(\.dismiss) private var dismiss

    // MARK: - Body

    var body: some View {
        ZStack {
            AppBackgroundView()

            if let viewModel {
                content(viewModel)
            }
        }
        .navigationTitle(Loc("Partija"))
        .navigationBarTitleDisplayMode(.inline)
        .toolbarBackground(Color.appBackground, for: .navigationBar)
        .toolbar {
            ToolbarItem(placement: .topBarTrailing) {
                // Predaja je jedini izlaz iz partije koja se ne moze dobiti.
                // Bez nje bi korak (a s njim i ostatak Puta) ostao zakljucan
                // dok korisnik ne matira racunar.
                if let viewModel, viewModel.canResign {
                    Button {
                        showResignConfirm = true
                    } label: {
                        Image(systemName: "flag.fill")
                            .foregroundStyle(DS.danger)
                    }
                }
            }
            ToolbarItem(placement: .topBarTrailing) {
                Button {
                    viewModel?.undo()
                } label: {
                    Image(systemName: "arrow.uturn.backward")
                }
                // `canUndo` je `false` kad je partija gotova, pa zavrsen korak
                // ne moze da se „odvrti" nazad u nezavrsenu partiju.
                .disabled(!(viewModel?.canUndo ?? false))
            }
        }
        .overlay(promotionOverlay)
        .alert(Loc("Predati partiju?"), isPresented: $showResignConfirm) {
            Button(Loc("Predaj"), role: .destructive) {
                viewModel?.resign()
            }
            Button(Loc("Otkaži"), role: .cancel) {}
        } message: {
            Text(Loc("Da li ste sigurni da želite da predate trenutnu partiju?"))
        }
        .onAppear(perform: start)
        // `isGameOver` je JEDINI izvor istine o kraju partije (`GameViewModel`);
        // ekran ga posmatra, ne izvodi ponovo.
        .onChange(of: viewModel?.isGameOver ?? false) { _, over in
            guard over else { return }
            completeStepOnce()
        }
    }

    // MARK: - Zivotni ciklus

    private func start() {
        if viewModel == nil {
            let model = GameViewModel(saveKey: GameViewModel.stepSaveKey(stepId))
            model.startStepGame(difficulty: difficulty, startFEN: startFEN)
            viewModel = model
        }
        pendingStepId = stepId
        // Branik za slucaj da je partija gotova pre nego sto `onChange` ima sta
        // da uporedi. `startStepGame` zavrsenu partiju ne nastavlja, pa je ovo
        // danas nedostizno — ali uslov zavrsetka stoji na oba puta do njega.
        completeStepOnce()
    }

    /// Korak se zavrsava kad partija dodje do kraja — BEZ OBZIRA NA ISHOD.
    /// Poraz i predaja zavrsavaju korak isto kao pobeda: spec trazi odigranu
    /// partiju, ne dobijenu.
    private func completeStepOnce() {
        guard let viewModel, viewModel.isGameOver, let id = pendingStepId else { return }
        pendingStepId = nil
        ProgressStore.shared.completeStep(id)
        // Bez dodatnog haptika: `GameViewModel` je za kraj partije vec ispalio
        // svoj (pobeda/poraz/remi) pre koji milisekundu.
        viewModel.clearStepSave()
    }

    // MARK: - Raspored

    private func content(_ viewModel: GameViewModel) -> some View {
        GeometryReader { geo in
            ScrollView {
                // Spacer-i na oba kraja + `minHeight`: kratak sadrzaj se
                // centrira po visini, dug normalno skroluje (isti obrazac kao
                // `StepPracticeView`).
                VStack(spacing: 0) {
                    Spacer(minLength: 0)
                    VStack(spacing: DS.Space.m) {
                        opponentCard(viewModel)
                        statusCard(viewModel)
                        board(viewModel, side: boardSide(in: geo.size))
                        footer(viewModel)
                    }
                    .frame(maxWidth: .infinity)
                    .padding(.horizontal, DS.Space.l)
                    .padding(.vertical, DS.Space.m)
                    Spacer(minLength: 0)
                }
                .frame(minHeight: geo.size.height)
            }
            .scrollBounceBehavior(.basedOnSize)
            .safeAreaPadding(.bottom, 24)
            .frame(width: geo.size.width, height: geo.size.height)
        }
    }

    /// Tabla mora da stane i kad je ekran polozen — bez granice po VISINI
    /// racun bi je vodio samo po sirini i u pejzazu bi progutala ceo ekran.
    private func boardSide(in size: CGSize) -> CGFloat {
        let byWidth = size.width - 2 * DS.Space.l
        let byHeight = size.height - 220
        return max(200, min(byWidth, byHeight, DS.maxBoardSide))
    }

    private func board(_ viewModel: GameViewModel, side: CGFloat) -> some View {
        BoardView(
            board:            viewModel.gameState.board,
            isFlipped:        viewModel.isFlipped,
            selectedPosition: viewModel.selectedPosition,
            legalMoves:       viewModel.legalMovesForSelected,
            lastMove:         viewModel.lastMove,
            animatingPiece:   viewModel.animatingPiece,
            flyingCapture:    viewModel.flyingCapture,
            playerColor:      viewModel.activePlayerColor,
            isPlayerTurn:     viewModel.isPlayerTurn,
            gameStatus:       viewModel.gameState.status,
            onTap:            { viewModel.tap(position: $0) }
        )
        .frame(width: side, height: side)
    }

    // MARK: - Protivnik

    private func opponentCard(_ viewModel: GameViewModel) -> some View {
        HStack(spacing: DS.Space.s) {
            Image(systemName: "cpu")
                .font(.dsBody)
                .foregroundStyle(DS.accent)
            Text(Loc("Računar"))
                .font(.dsBody)
                .foregroundStyle(DS.ink)
            Text(viewModel.difficulty.label)
                .font(.dsCaption)
                .foregroundStyle(DS.inkMuted)
            Spacer(minLength: 0)
            if viewModel.isThinking {
                ThinkingIndicator(isActive: true)
            }
        }
        .padding(.horizontal, DS.Space.m)
        .padding(.vertical, DS.Space.m)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(DS.surface, in: RoundedRectangle(cornerRadius: DS.Radius.m))
        .overlay(RoundedRectangle(cornerRadius: DS.Radius.m).strokeBorder(DS.line, lineWidth: 1))
        .accessibilityElement(children: .combine)
    }

    // MARK: - Status

    private func statusCard(_ viewModel: GameViewModel) -> some View {
        let style = statusStyle(viewModel)
        return HStack(spacing: DS.Space.s) {
            Image(systemName: style.icon)
                .font(.dsBody)
                .foregroundStyle(style.color)
            Text(viewModel.statusMessage)
                .font(.dsBody)
                .foregroundStyle(style.color)
                .animation(.default, value: viewModel.statusMessage)
            Spacer(minLength: 0)
        }
        .padding(.horizontal, DS.Space.m)
        .padding(.vertical, DS.Space.m)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(style.color.opacity(0.12), in: RoundedRectangle(cornerRadius: DS.Radius.m))
        .overlay(RoundedRectangle(cornerRadius: DS.Radius.m)
            .strokeBorder(style.color.opacity(0.3), lineWidth: 1))
    }

    /// Boja nosi ISHOD, ne ukras: zeleno kad je korisnik dobio, crveno kad je
    /// izgubio ili predao, zuto za remi i sah.
    private func statusStyle(_ viewModel: GameViewModel) -> (icon: String, color: Color) {
        switch viewModel.gameState.status {
        case .playing:
            return ("info.circle", DS.ink)
        case .check:
            return ("exclamationmark.triangle.fill", DS.warning)
        case .draw:
            return ("equal.circle.fill", DS.warning)
        case .checkmate(let loser), .resigned(let loser):
            return loser == viewModel.playerColor
                ? ("xmark.circle.fill", DS.danger)
                : ("checkmark.seal.fill", DS.success)
        }
    }

    // MARK: - Zavrsetak koraka

    @ViewBuilder
    private func footer(_ viewModel: GameViewModel) -> some View {
        if viewModel.isGameOver {
            VStack(spacing: DS.Space.m) {
                HStack(spacing: DS.Space.s) {
                    Image(systemName: "checkmark.circle.fill")
                        .foregroundStyle(DS.success)
                    Text(Loc("Korak je završen"))
                        .foregroundStyle(DS.ink)
                }
                .font(.dsHeading)

                Button {
                    dismiss()
                } label: {
                    Text(Loc("Nazad na Put"))
                        .font(.dsBody.weight(.semibold))
                        .foregroundStyle(DS.onAccent)
                        .frame(maxWidth: .infinity)
                        .padding(.vertical, DS.Space.m)
                        .background(DS.accent, in: RoundedRectangle(cornerRadius: DS.Radius.m))
                }
                .buttonStyle(.plain)
            }
            .padding(.top, DS.Space.s)
        } else {
            // Uslov zavrsetka se kaze unapred: korisnik koji ocekuje da mora da
            // POBEDI odustao bi od izgubljene partije umesto da je odigra.
            Text(Loc("Korak se završava kad partija dođe do kraja."))
                .font(.dsCaption)
                .foregroundStyle(DS.inkMuted)
                .multilineTextAlignment(.center)
                .frame(maxWidth: .infinity)
        }
    }

    // MARK: - Promocija

    @ViewBuilder
    private var promotionOverlay: some View {
        if let viewModel, viewModel.showPromotion {
            PromotionOverlay(color: viewModel.playerColor) { type in
                viewModel.confirmPromotion(type)
            }
        }
    }
}

File: ChesskoAndroid/app/src/main/java/com/veljkoni/chessko/ui/StepGameView.kt
package com.veljkoni.chessko.ui

import android.app.Application
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.Undo
import androidx.compose.material.icons.filled.Computer
import androidx.compose.material.icons.filled.Flag
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.window.Dialog
import androidx.compose.ui.window.DialogProperties
import com.veljkoni.chessko.logic.GameDifficulty
import com.veljkoni.chessko.logic.ProgressStore
import com.veljkoni.chessko.logic.SettingsManager
import com.veljkoni.chessko.logic.loc
import com.veljkoni.chessko.ui.theme.DS
import com.veljkoni.chessko.ui.theme.Type
import com.veljkoni.chessko.viewmodels.GameViewModel

// MARK: - Pokretac koraka `game` (Faza 6c, Task 7)
//
// Partija protiv racunara sa tezinom (i opciono startnom pozicijom) iz
// koraka. Ceo tok partije -- tap, legalnost, AI, zvuk, haptika, promocija,
// kraj partije -- vec zivi u `GameViewModel`-u i deli se sa tabom Igra; ovaj
// ekran dodaje samo ono sto korak ima a slobodna partija nema: zavrsetak
// KORAKA.
//
// Tri odluke koje ekran nosi:
// 1. SOPSTVENI `GameViewModel`, sa sopstvenim slotom za cuvanje
//    (`GameViewModel.stepSaveKey(stepId)`). Deljeni primerak (ili zajednicki
//    kljuc) bi ulaskom u korak ucitao/pregazio partiju koju korisnik ima u
//    toku na tabu Igra -- isti razlog zbog kog `StepPracticeView` drzi
//    sopstveni `PuzzleViewModel`.
// 2. Korak se zavrsava kad partija dodje do KRAJA, BEZ OBZIRA NA ISHOD --
//    spec trazi "partija odigrana do kraja", ne pobedu. Predaja je takodje
//    kraj.
// 3. Upis koraka je gejtovan na `pendingStepId`, da se ne bi desio dvaput za
//    isti ulazak (npr. ako se status pozicije ponovo izracuna dok je partija
//    vec gotova).
@Composable
fun StepGameView(difficulty: String, startFEN: String?, stepId: String, onClose: () -> Unit) {
    val context = LocalContext.current
    val app = context.applicationContext as Application
    val progressStore = remember { ProgressStore.getInstance(app) }

    // Sirov string iz kurikuluma ("beginner", "easy", ...) -- `Curriculum.parse`
    // ga vec proverava protiv `KNOWN_DIFFICULTIES`, pa je ovo mapiranje totalno
    // u praksi; fallback na `MEDIUM` postoji samo da ekran nikad ne padne.
    val gameDifficulty = remember(difficulty) {
        GameDifficulty.values().find { it.name.equals(difficulty, ignoreCase = true) }
            ?: GameDifficulty.MEDIUM
    }

    // Sopstveni model po `stepId`: novi korak dobija nov primerak, a s njim i
    // nov slot za cuvanje (`stepSaveKey`).
    val viewModel = remember(stepId) { GameViewModel(app, GameViewModel.stepSaveKey(stepId)) }

    // `viewModel` je napravljen kroz `remember`, NE kroz `ViewModelStore`, pa mu se
    // `onCleared()` NIKAD ne izvrsi -- `SoundManager` (nije singleton, svaki
    // `GameViewModel` pravi SVOJ `SoundPool`) bi bez ovoga procureo na svaki izlazak
    // iz koraka. Kljuc je `stepId`, isto kao `remember` iznad: ako se `stepId`
    // promeni dok je ekran ziv, STARI model se oslobadja pre nego sto se efekat
    // za novi instalira, ne tek kad ceo ekran nestane. Isti obrazac kao
    // `ChessClockView.kt` (`DisposableEffect` + `release()`).
    DisposableEffect(stepId) {
        onDispose { viewModel.releaseSounds() }
    }

    // `null` znaci "vec upisan" -- zavrsetak se okida TACNO jednom po ulasku,
    // ne na svako ponovno izracunavanje statusa partije.
    var pendingStepId by remember(stepId) { mutableStateOf<String?>(stepId) }

    var showResignConfirm by remember { mutableStateOf(false) }
    var showAnalysis by remember { mutableStateOf(false) }

    LaunchedEffect(stepId) {
        viewModel.startStepGame(gameDifficulty, startFEN)
        // Branik za slucaj da je partija gotova pre nego sto donji efekat ima
        // sta da uporedi. `startStepGame` zavrsenu partiju ne nastavlja (resetuje
        // je), pa je ovo danas nedostizno -- ali uslov zavrsetka stoji na oba
        // puta do njega, isto kao iOS `start()`/`completeStepOnce()`.
        if (viewModel.isGameOver && pendingStepId == stepId) {
            pendingStepId = null
            progressStore.completeStep(stepId)
            GameViewModel.clearStepSave(app, stepId)
        }
    }

    // Reaguje na SVAKU promenu statusa partije (potez, mat, remi, predaja).
    LaunchedEffect(viewModel.gameState.status) {
        if (viewModel.isGameOver && pendingStepId == stepId) {
            pendingStepId = null
            progressStore.completeStep(stepId)
            GameViewModel.clearStepSave(app, stepId)
        }
    }

    Column(
        modifier = Modifier.fillMaxSize().padding(16.dp),
        verticalArrangement = Arrangement.spacedBy(14.dp)
    ) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.SpaceBetween,
            modifier = Modifier.fillMaxWidth()
        ) {
            Text(
                text = loc("Partija"),
                color = DS.ink,
                // Zaglavlje ekrana koraka -> `heading`, isto kao `StepPracticeView`.
                style = Type.heading,
                fontWeight = FontWeight.Bold
            )
            Row(horizontalArrangement = Arrangement.spacedBy(4.dp)) {
                Button(
                    onClick = { viewModel.undo() },
                    enabled = viewModel.canUndo,
                    colors = ButtonDefaults.buttonColors(
                        containerColor = Color.Transparent,
                        contentColor = DS.ink,
                        // `disabledContainerColor` se NE sme izostaviti: M3
                        // podrazumevano daje `onSurface` @12%, pa afordansa cita
                        // naopako — onemoguceno dugme postane popunjena siva pilula,
                        // a omoguceno goli glif.
                        disabledContainerColor = Color.Transparent,
                        // Ranije je i ovo bilo `DS.inkMuted`, identicno sa
                        // `contentColor` — dugme je izgledalo isto u oba stanja.
                        // Ostatak grane razlikuje stanja bas ovim parom
                        // (`MainActivity.kt:1239-1241`, `:1275-1277`,
                        // `PuzzleView.kt:454-457`): `DS.ink` ukljuceno, `DS.inkMuted`
                        // iskljuceno. Podloga je `DS.ground`; izmereno
                        // `ink`/`ground` = 15,72 (svetla) / 16,15 (tamna),
                        // `inkMuted`/`ground` = 4,36 / 5,94.
                        disabledContentColor = DS.inkMuted
                    )
                ) {
                    // Faza 8, talas ispravki: `Text("↩︎")` -> `Icon`. Glif je
                    // prezivEO celu fazu jer je provera obima grepovala samo
                    // emoji blokove (`1F300–1FAFF`, `2600–27BF`, `2B00–2BFF`),
                    // a `U+21A9` je u bloku strelica. Dugme nema tekst — ikona
                    // stoji SAMA, pa dobija opis; isti kljuc i ista radnja kao
                    // `Undo` u `MainActivity.kt` (gde je dekorativna, jer je
                    // tamo uz vidljiv `Text`). `AutoMirrored` je bitan: glif
                    // se u RTL rasporedu mora okrenuti, sto goli `Text` nije
                    // radio. Boja se nasledjuje iz `ButtonDefaults` iznad
                    // (`DS.ink` / `DS.inkMuted`), par vec izmeren u komentaru.
                    Icon(
                        imageVector = Icons.AutoMirrored.Filled.Undo,
                        contentDescription = loc("Vrati potez"),
                        modifier = Modifier.size(20.dp)
                    )
                }
                if (viewModel.canResign) {
                    Button(
                        onClick = { showResignConfirm = true },
                        colors = ButtonDefaults.buttonColors(
                            containerColor = Color.Transparent,
                            contentColor = DS.danger
                        )
                    ) {
                        // Faza 8, Task 3: emoji -> Icon. Dugme nema tekst pored
                        // sebe (za razliku od "Zatvori" odmah desno) — opis je
                        // neophodan. Isti kljuc kao `Flag` u `MainActivity.kt`
                        // (Faza 8, Task 2), ista radnja.
                        Icon(
                            imageVector = Icons.Default.Flag,
                            contentDescription = loc("Predaj partiju"),
                            modifier = Modifier.size(20.dp)
                        )
                    }
                }
                Button(
                    onClick = onClose,
                    colors = ButtonDefaults.buttonColors(
                        containerColor = Color.Transparent,
                        contentColor = DS.inkMuted
                    )
                ) {
                    Text(text = loc("Zatvori"))
                }
            }
        }

        OpponentCard(viewModel)
        StatusCard(message = getStatusMessage(viewModel))

        val settings = remember { SettingsManager.getInstance(context) }
        val activeTheme = remember(settings.boardTheme) {
            BoardTheme.values().find { it.rawValue == settings.boardTheme } ?: BoardTheme.CLASSIC
        }
        val activeStyle = remember(settings.pieceStyle) {
            PieceStyle.values().find { it.rawValue == settings.pieceStyle } ?: PieceStyle.CLASSIC
        }

        Box(modifier = Modifier.weight(1f), contentAlignment = Alignment.Center) {
            BoardView(
                board = viewModel.gameState.board,
                isFlipped = viewModel.isFlipped,
                selectedPosition = viewModel.selectedPosition,
                legalMoves = viewModel.legalMovesForSelected,
                lastMove = viewModel.lastMove,
                gameStatus = viewModel.gameState.status,
                boardTheme = activeTheme,
                pieceStyle = activeStyle,
                showCoordinates = settings.showCoordinates,
                showLastMoveHighlight = settings.showLastMoveHighlight,
                showLegalMoves = settings.showLegalMoves,
                onTap = { pos -> viewModel.tap(pos) },
                modifier = Modifier.widthIn(max = DS.maxBoardSide)
            )
        }

        StepGameFooter(
            viewModel = viewModel,
            isGameOver = viewModel.isGameOver,
            onClose = onClose,
            onAnalyze = { showAnalysis = true }
        )
    }

    if (viewModel.showPromotion) {
        PromotionOverlay(viewModel)
    }

    // Korak se upisuje kao zavrsen PRE ovoga (vidi LaunchedEffect-e iznad,
    // oba gejtovana na `isGameOver && pendingStepId == stepId`), pa analiza
    // NIKAD ne stoji na putu zavrsetka koraka -- ako motor nije dostupan,
    // korak je vec zavrsen, samo dugme ostaje bez efekta.
    if (showAnalysis) {
        Dialog(
            onDismissRequest = { showAnalysis = false },
            properties = DialogProperties(usePlatformDefaultWidth = false)
        ) {
            AnalysisView(viewModel = viewModel, onClose = { showAnalysis = false })
        }
    }

    if (showResignConfirm) {
        AlertDialog(
            onDismissRequest = { showResignConfirm = false },
            title = { Text(text = loc("Predaja partije"), color = DS.ink, fontWeight = FontWeight.Bold) },
            // 0.8f je van tacnog opsega inkMuted (0.5-0.75f) iz tabele, ali je
            // ovo bas telo/supporting tekst dijaloga -- isti M3 slot
            // (`DialogTokens.SupportingTextColor`) koji `Theme.kt` vec mapira na
            // `onSurfaceVariant` = `inkMuted` po podrazumevanoj vrednosti.
            text = { Text(text = loc("Da li ste sigurni da želite da predate partiju?"), color = DS.inkMuted) },
            confirmButton = {
                TextButton(onClick = {
                    viewModel.resign()
                    showResignConfirm = false
                }) {
                    Text(text = loc("Predaj"), color = DS.danger, fontWeight = FontWeight.Bold)
                }
            },
            dismissButton = {
                TextButton(onClick = { showResignConfirm = false }) {
                    Text(text = loc("Nastavi igru"), color = DS.inkMuted)
                }
            },
            containerColor = DS.surface,
            shape = RoundedCornerShape(16.dp)
        )
    }
}

@Composable
private fun OpponentCard(viewModel: GameViewModel) {
    Row(
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(8.dp),
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(12.dp))
            .background(DS.fill)
            .padding(horizontal = 14.dp, vertical = 12.dp)
    ) {
        // Faza 8, Task 3: emoji -> Icon. Ikona stoji NEPOSREDNO uz vidljiv
        // `Text(loc("Računar"))` -- isti obrazac kao `Computer`/`People` u
        // `MainActivity.kt` (Faza 8, Task 2, ispravljeno u progress.md nakon
        // sto je prvobitni kljuc "Protivnik računar" bio dodeljen ovom istom
        // slucaju): dekorativna, BEZ kljuca (ne samo `null` opis vec i BEZ
        // dodatog `Loc.kt` unosa) -- ovaj korak, za razliku od slobodne
        // partije na tabu Igra, nema prekidac Racunar/Prijatelj, pa se ikona
        // ovde nikad ne menja u `People`.
        Icon(
            imageVector = Icons.Default.Computer,
            contentDescription = null,
            tint = DS.ink,
            modifier = Modifier.size(16.dp)
        )
        // Kartica protivnika je paritet sa iOS-om
        // (`Chessko/Views/StepGameView.swift:199-205`): naziv `.dsBody`, tezina
        // `.dsCaption`. `fontWeight = Medium` ostaje -- iOS ga nema, ali brisanje
        // bi bila promena izgleda koju ovaj task ne trazi.
        Text(text = loc("Računar"), color = DS.ink, style = Type.body, fontWeight = FontWeight.Medium)
        Text(text = viewModel.difficulty.label, color = DS.inkMuted, style = Type.caption)
        Spacer(modifier = Modifier.weight(1f))
        if (viewModel.isThinking) {
            Text(text = loc("Računar razmišlja..."), color = DS.inkMuted, style = Type.caption)
        }
    }
}

@Composable
private fun StatusCard(message: String) {
    Row(
        verticalAlignment = Alignment.CenterVertically,
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(12.dp))
            .background(DS.fill)
            .padding(horizontal = 14.dp, vertical = 12.dp)
    ) {
        // iOS: `.dsBody` (`Chessko/Views/StepGameView.swift:229`).
        Text(text = message, color = DS.ink, style = Type.body)
    }
}

@Composable
private fun StepGameFooter(
    viewModel: GameViewModel,
    isGameOver: Boolean,
    onClose: () -> Unit,
    onAnalyze: () -> Unit
) {
    if (isGameOver) {
        Column(
            modifier = Modifier.fillMaxWidth(),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(10.dp)
        ) {
            Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(6.dp)) {
                Box(modifier = Modifier.size(8.dp).clip(CircleShape).background(DS.success))
                Text(text = loc("Korak je završen"), color = DS.ink, fontWeight = FontWeight.SemiBold)
            }
            // Korak je vec upisan kao zavrsen (vidi komentar na mestu poziva) --
            // ovo dugme je cisto dodatna informacija, ne uslov za nastavak Puta.
            AnalysisButton(viewModel = viewModel, onClick = onAnalyze)
            Button(
                onClick = onClose,
                modifier = Modifier.fillMaxWidth(),
                colors = ButtonDefaults.buttonColors(containerColor = DS.accent, contentColor = DS.onAccent)
            ) {
                Text(text = loc("Nazad na Put"), fontWeight = FontWeight.Bold)
            }
        }
    } else {
        Text(
            text = loc("Korak se završava kad partija dođe do kraja."),
            color = DS.inkMuted,
            // iOS: `.dsCaption` (`Chessko/Views/StepGameView.swift:290`).
            style = Type.caption,
            modifier = Modifier.fillMaxWidth()
        )
    }
}
